### PR TITLE
Optimize GDV checks for objects with known type

### PR DESCRIPTION
**Pull request by EgorBo**
Optimize `*obj == const_class` GDV's guards if we can get the exact type of `obj`. This is needed for https://github.com/dotnet/runtime/issues/84659

Codegen improvement example:
```csharp
string Test(byte[] data) => Encoding.UTF8.GetString(data);
```
Was:
```asm
; Assembly listing for method Prog:Test(ubyte[]):System.String:this
; Tier-1 compilation
; 1 inlinees with PGO data; 1 single block inlinees; 0 inlinees without PGO data
       sub      rsp, 40
       mov      rcx, 0x1C4C0C009F0      ; const ptr
       mov      rcx, gword ptr [rcx]
       mov      rax, 0x7FFE3DD8A348      ; System.Text.UTF8Encoding+UTF8EncodingSealed
       cmp      qword ptr [rcx], rax ;;; <---- redundant check, obj is already of UTF8EncodingSealed
       jne      SHORT G_M52604_IG07
       test     rdx, rdx
       je       SHORT G_M52604_IG04
       cmp      dword ptr [rdx+08H], 32
       jg       SHORT G_M52604_IG04
       call     [System.Text.UTF8Encoding+UTF8EncodingSealed:GetStringForSmallInput(ubyte[]):System.String:this]
       jmp      SHORT G_M52604_IG05
G_M52604_IG04:  
       call     [System.Text.Encoding:GetString(ubyte[]):System.String:this]
G_M52604_IG05:  
       nop      
       add      rsp, 40
       ret      
G_M52604_IG07:  
       call     [System.Text.UTF8Encoding+UTF8EncodingSealed:GetString(ubyte[]):System.String:this]
       jmp      SHORT G_M52604_IG05
; Total bytes of code 71
```
Now:
```asm
; Assembly listing for method Prog:Test(ubyte[]):System.String:this
; Tier-1 compilation
; optimized code
; 1 inlinees with PGO data; 1 single block inlinees; 0 inlinees without PGO data
       sub      rsp, 40
       mov      rcx, 0x1B4148009F0      
       mov      rcx, gword ptr [rcx]
       test     rdx, rdx
       je       SHORT G_M52604_IG04
       cmp      dword ptr [rdx+08H], 32
       jg       SHORT G_M52604_IG04
       call     [System.Text.UTF8Encoding+UTF8EncodingSealed:GetStringForSmallInput(ubyte[]):System.String:this]
       jmp      SHORT G_M52604_IG05
G_M52604_IG04:  
       call     [System.Text.Encoding:GetString(ubyte[]):System.String:this]
G_M52604_IG05:  
       nop      
       add      rsp, 40
       ret      
; Total bytes of code 48
```

### CODE REVIEW

**Comment by EgorBo** [comment]
PTAL @AndyAyersMS - I wanted to land this fix in the past and gave up but now I found a real use-case

**Comment by AndyAyersMS** [comment]
Seems like something we already should have been able to clean up.

What is it we know when we do this new type compare fold that we don't know later on (say in value numbering)?

**Comment by EgorBo** [comment]
> Seems like something we already should have been able to clean up.
> 
> What is it we know when we do this new type compare fold that we don't know later on (say in value numbering)?

I don't think we have anything like that in VN, you probably confuse it with your https://github.com/dotnet/runtime/pull/72136 which is different. I could move this opt to VN as well but all the cases I found looked like can be handled earlier in Morph. Essentially, it happens when we have something like this:
```
Foo().Bar();
```
when we look at `Bar` virtual call we mark it as GDV candidate because we don't know exact type of `Foo()`. Then later when we inline `Foo()` and get something with exact type we ended up with that redundant GDV check. So it sort of inlining-ordering issue that we clean up here.

If you want to repro it locally:
```csharp
public class ClassA
{
    public virtual int GetVal() => 42;
}

public sealed class ClassB : ClassA
{
    public override int GetVal() => 43;
}

class Prog
{
    static readonly ClassB s_default = new ClassB();

    static ClassA Default => s_default;

    static void Main()
    {
        for (int i = 0; i < 100; i++)
        {
            Test();
            Thread.Sleep(16);
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Test() => Default.GetVal();
}
```
Run with PGO

**Comment by AndyAyersMS** [comment]
With your changes is this firing in `LateDevirtualization`?

> I don't think we have anything like that in VN, you probably confuse it with your https://github.com/dotnet/runtime/pull/72136 which is different.

Is what you're saying is that VN doesn't know how to handle method table loads from these "const ptrs"?

I'm ok fixing it in morph but remember that morph is purely local and if a known value arrives at the comparison via some data flow morph will never see it. So consider fixing it in VN as well.

**Comment by EgorBo** [comment]
@AndyAyersMS I've just merged VN impl 🙂 It turned out VN knew it only if it could find `VNF_JitNew` func, so basically track type only for allocations

**Comment by EgorBo** [comment]
It seems it found quite a few cases for my app running locally with PGO.

**Review by AndyAyersMS — APPROVED** [review]
LGTM, but I am ok also having it checked earlier too, like you were doing.

Usually, the earlier the better for these things.

**Comment by EgorBo** [comment]
Some [nice diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=236185&view=ms.vss-build-web.run-extensions-tab) in `benchmarks.run.pgo` collection (thanks for adding it!). We don't have apsnet one but presumably it should also hit them. I didn't see any diff improvements from the change in `morph` so I removed it. 

Although, it seems that NativeAOT is not happy

**Comment by AndyAyersMS** [comment]
I just redid the ASP.NET collection but maybe another GUID change came in that I missed. Will start it going again.

**Comment by AndyAyersMS** [comment]
> I just redid the ASP.NET collection but maybe another GUID change came in that I missed. Will start it going again.

The new collection just uploaded.

**Comment by EgorBo** [comment]
> > I just redid the ASP.NET collection but maybe another GUID change came in that I missed. Will start it going again.
> 
> The new collection just uploaded.

Thanks! `-15237 bytes` on that one.

**Comment by EgorBo** [comment]
hm.. NativeAOT struggled from generic types on that path, e.g. `System.Collections.Generic.HashSet`1[System.__Canon]` so I added a check to trim them

**Comment by EgorBo** [comment]
No idea why NativeAOT complains 🙁

File: src/coreclr/jit/valuenum.cpp [hunk 1]
```diff
@@ -10617,13 +10617,16 @@ void Compiler::fgValueNumberTree(GenTree* tree)
...
+                            CORINFO_INLINE_TYPECHECK_PASS)
                         {
-                            JITDUMP("IND(obj) is actually a class handle for %s\n", eeGetClassName(handle));
-                            ValueNum handleVN = vnStore->VNForHandle((ssize_t)handle, GTF_ICON_CLASS_HDL);
-                            tree->gtVNPair    = vnStore->VNPWithExc(ValueNumPair(handleVN, handleVN), addrXvnp);
-                            returnsTypeHandle = true;
+                            if (info.compCompHnd->compareTypesForEquality(handle, handle) == TypeCompareState::Must)
+                            {
+                                JITDUMP("IND(obj) is actually a class handle for %s\n", eeGetClassName(handle));
+                                ValueNum handleVN = vnStore->VNForHandle((ssize_t)handle, GTF_ICON_CLASS_HDL);
```
Comment (SingleAccretion) [inline src/coreclr/jit/valuenum.cpp]: > `ValueNum handleVN = vnStore->VNForHandle((ssize_t)handle, GTF_ICON_CLASS_HDL);`

This assumes the compile-time handle is the same as the runtime one, which is not true for AOT. The handle needs to be "embedded".

**Comment by EgorBo** [comment]
Addressed

File: src/coreclr/jit/valuenum.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/valuenum.cpp]: @SingleAccretion thanks!! I keep forgetting about it 🙂 does it look good now? I see that there is a cache of handles in VN but in my case it's not guaranteed that the handle will be there + this path is taken rarely in AOT

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -10606,23 +10606,46 @@ void Compiler::fgValueNumberTree(GenTree* tree)
...
-                        tree->gtVNPair =
-                            vnStore->VNPWithExc(ValueNumPair(funcApp.m_args[0], funcApp.m_args[0]), addrXvnp);
-                        wasNewobj = true;
+                        info.compCompHnd->classMustBeLoadedBeforeCodeIsRun(handle);
+                        JITDUMP("IND(obj) is actually a class handle for %s\n", eeGetClassName(handle));
+                        void* pEmbedClsHnd;
+                        void* embedClsHnd = (void*)info.compCompHnd->embedClassHandle(handle, &pEmbedClsHnd);
+                        if (pEmbedClsHnd == nullptr)
+                        {
+                            // Skip indirect handles for now since this path is mostly for PGO scenarios
```
Comment (SingleAccretion) [inline src/coreclr/jit/valuenum.cpp:10626]: Note: indirect handles could be handled by essentially switching out the address for the indirection to be `pEmbedClsHnd`.

File: src/coreclr/jit/valuenum.cpp [hunk 3]
```diff
@@ -10606,23 +10606,46 @@ void Compiler::fgValueNumberTree(GenTree* tree)
...
+                            assert(embedClsHnd != nullptr);
+                            ssize_t  cnsHandle = (ssize_t)(pEmbedClsHnd != nullptr ? pEmbedClsHnd : embedClsHnd);
+                            ValueNum handleVN  = vnStore->VNForHandle(cnsHandle, GTF_ICON_CLASS_HDL);
```
Comment (SingleAccretion) [inline src/coreclr/jit/valuenum.cpp:10629]: `pEmbedClsHnd != nullptr` is dead code now.

**Review by SingleAccretion — APPROVED** [review]

File: src/coreclr/jit/valuenum.cpp [hunk 4]
```diff
@@ -10606,23 +10606,45 @@ void Compiler::fgValueNumberTree(GenTree* tree)
...
+                    //
+                    bool                 isExact   = false;
+                    bool                 isNonNull = false;
+                    CORINFO_CLASS_HANDLE handle    = gtGetClassHandle(addr, &isExact, &isNonNull);
+                    if (isExact && (handle != NO_CLASS_HANDLE))
                     {
-                        tree->gtVNPair =
-                            vnStore->VNPWithExc(ValueNumPair(funcApp.m_args[0], funcApp.m_args[0]), addrXvnp);
-                        wasNewobj = true;
+                        info.compCompHnd->classMustBeLoadedBeforeCodeIsRun(handle);
```
Comment (jkotas) [inline src/coreclr/jit/valuenum.cpp:10620]:
```suggestion
```
This should not be needed. `embedClassHandle` takes care of that.

**Comment by EgorBo** [comment]
I'm struggling with a NativeAOT assert (happens when ILC compiles crossgen2 project) - it happens inside `embedClassHandle` EE API Jit calls:

image

where the values in the assert are:
```
{[S.P.CoreLib]System.Action`1<System.Collections.Generic.List`1<ILCompiler.DependencyAnalysisFramework.DependencyNodeCore`1<System.__Canon>>>}

==

{[S.P.CoreLib]System.Action`1<System.__Canon>}
````

@MichalStrehovsky @jkotas does it tell you anything, I am allowed to use shared types at all in that API?

**Comment by jkotas** [comment]
The assert is complaining about directly embedding type handle that requires dictionary lookup. It is not right in any configs. The types that require dictionary lookup should be never embedded directly. You can try to check `CORINFO_FLG_SHAREDINST` to see whether the type requires dictionary lookup.

**Comment by EgorBo** [comment]
> The assert is complaining about directly embedding type handle that requires dictionary lookup. It is not right in any configs. The types that require dictionary lookup should be never embedded directly. You can try to check `CORINFO_FLG_SHAREDINST` to see whether the type requires dictionary lookup.

Thanks, that makes perfect sense

**Comment by EgorBo** [comment]
The check has fixed the assert and the diffs are back to normal: https://dev.azure.com/dnceng-public/public/_build/results?buildId=248263&view=ms.vss-build-web.run-extensions-tab

**Comment by EgorBo** [comment]
all CI jobs passed except license/CLA 😢

**Pull request merged**